File: TreeMachine.Pro_Cpp/TreeMachine.Pro.Tests/Sources/TreeMachine/Tests_00.cpp
#include "gtest/gtest.h"
#include "Tree.h"

namespace TreeMachine {
    using namespace std;

    //     TEST(Tests_00, Test) { // NOLINT
    // #if defined(__clang__) && !defined(_MSC_VER)
    //         SUCCEED() << "Compiler: Clang" << endl;
    // #elif defined(__clang__) && defined(_MSC_VER)
    //         SUCCEED() << "Compiler: Clang-cl (Clang with MSVC compatibility)" << endl;
    // #elif defined(_MSC_VER)
    //         SUCCEED() << "Compiler: MSVC" << endl;
    // #elif defined(__GNUC__)
    //         SUCCEED() << "Compiler: GCC" << endl;
    // #else
    //         SUCCEED() << "Compiler: Unknown" << endl;
    // #endif
    //     }

    TEST(Tests_00, Test_00) { // NOLINT
        auto *const tree = new Tree();
        auto *const root = new Root();
        auto *const a = new A();
        auto *const b = new B();

        for (const auto *const node : vector<Node *>{root, a, b}) {
            EXPECT_NE(node, nullptr);
            EXPECT_EQ(node->Tree(), nullptr);
            EXPECT_EQ(node->IsRoot(), true);
            EXPECT_EQ(node->Root(), node);
            EXPECT_EQ(node->Parent(), nullptr);
            EXPECT_EQ(node->Ancestors().size(), 0);
            EXPECT_EQ(node->AncestorsAndSelf().size(), 1);
            EXPECT_EQ(node->Activity(), Node::Activity_::Inactive);
            EXPECT_EQ(node->Children().size(), 0);
            EXPECT_EQ(node->Descendants().size(), 0);
            EXPECT_EQ(node->DescendantsAndSelf().size(), 1);
        }

        {
            // tree.AddRoot root
            tree->AddRoot(root, nullptr);
            EXPECT_NE(tree->Root(), nullptr);
            EXPECT_EQ(tree->Root()->Tree(), tree);
            EXPECT_EQ(tree->Root()->IsRoot(), true);
            EXPECT_EQ(tree->Root()->Root(), root);
            EXPECT_EQ(tree->Root()->Parent(), nullptr);
            EXPECT_EQ(tree->Root()->Ancestors().size(), 0);
            EXPECT_EQ(tree->Root()->AncestorsAndSelf().size(), 1);
            EXPECT_EQ(tree->Root()->Activity(), Node::Activity_::Active);
            EXPECT_EQ(tree->Root()->Children().size(), 0);
            EXPECT_EQ(tree->Root()->Descendants().size(), 0);
            EXPECT_EQ(tree->Root()->DescendantsAndSelf().size(), 1);
        }
        {
            // tree.Root.AddChildren a, b
            tree->Root()->AddChildren(vector<Node *>{a, b}, nullptr);
            EXPECT_NE(tree->Root(), nullptr);
            EXPECT_EQ(tree->Root()->Tree(), tree);
            EXPECT_EQ(tree->Root()->IsRoot(), true);
            EXPECT_EQ(tree->Root()->Root(), root);
            EXPECT_EQ(tree->Root()->Parent(), nullptr);
            EXPECT_EQ(tree->Root()->Ancestors().size(), 0);
            EXPECT_EQ(tree->Root()->AncestorsAndSelf().size(), 1);
            EXPECT_EQ(tree->Root()->Activity(), Node::Activity_::Active);
            EXPECT_EQ(tree->Root()->Children().size(), 2);
            EXPECT_EQ(tree->Root()->Descendants().size(), 2);
            EXPECT_EQ(tree->Root()->DescendantsAndSelf().size(), 3);
            for (const auto *const child : tree->Root()->Children()) {
                EXPECT_EQ(child->Tree(), tree);
                EXPECT_EQ(child->IsRoot(), false);
                EXPECT_EQ(child->Root(), root);
                EXPECT_EQ(child->Parent(), root);
                EXPECT_EQ(child->Ancestors().size(), 1);
                EXPECT_EQ(child->AncestorsAndSelf().size(), 2);
                EXPECT_EQ(child->Activity(), Node::Activity_::Active);
                EXPECT_EQ(child->Children().size(), 0);
                EXPECT_EQ(child->Descendants().size(), 0);
                EXPECT_EQ(child->DescendantsAndSelf().size(), 1);
            }
        }
        {
            // tree.Root.RemoveChildren
            tree->Root()->RemoveChildren(nullptr, [](auto *const child, [[maybe_unused]] const auto arg) { delete child; });
            EXPECT_NE(tree->Root(), nullptr);
            EXPECT_EQ(tree->Root()->Tree(), tree);
            EXPECT_EQ(tree->Root()->IsRoot(), true);
            EXPECT_EQ(tree->Root()->Root(), root);
            EXPECT_EQ(tree->Root()->Parent(), nullptr);
            EXPECT_EQ(tree->Root()->Ancestors().size(), 0);
            EXPECT_EQ(tree->Root()->AncestorsAndSelf().size(), 1);
            EXPECT_EQ(tree->Root()->Activity(), Node::Activity_::Active);
            EXPECT_EQ(tree->Root()->Children().size(), 0);
            EXPECT_EQ(tree->Root()->Descendants().size(), 0);
            EXPECT_EQ(tree->Root()->DescendantsAndSelf().size(), 1);
        }
        {
            // tree.RemoveRoot
            tree->RemoveRoot(nullptr, [](auto *const root, [[maybe_unused]] const auto arg) { delete root; }); // NOLINT
            EXPECT_EQ(tree->Root(), nullptr);
        }

        delete tree;
    }

    TEST(Tests_00, Test_01) { // NOLINT
        auto *const tree = new Tree();
        auto *const root = new Root();

        for (const auto *const node : vector<Node *>{root}) {
            EXPECT_NE(node, nullptr);
            EXPECT_EQ(node->Tree(), nullptr);
            EXPECT_EQ(node->IsRoot(), true);
            EXPECT_EQ(node->Root(), node);
            EXPECT_EQ(node->Parent(), nullptr);
            EXPECT_EQ(node->Ancestors().size(), 0);
            EXPECT_EQ(node->AncestorsAndSelf().size(), 1);
            EXPECT_EQ(node->Activity(), Node::Activity_::Inactive);
            EXPECT_EQ(node->Children().size(), 0);
            EXPECT_EQ(node->Descendants().size(), 0);
            EXPECT_EQ(node->DescendantsAndSelf().size(), 1);
        }

        {
            // tree.AddRoot
            tree->AddRoot(root, nullptr);
            EXPECT_NE(tree->Root(), nullptr);
            EXPECT_EQ(tree->Root()->Tree(), tree);
            EXPECT_EQ(tree->Root()->IsRoot(), true);
            EXPECT_EQ(tree->Root()->Root(), root);
            EXPECT_EQ(tree->Root()->Parent(), nullptr);
            EXPECT_EQ(tree->Root()->Ancestors().size(), 0);
            EXPECT_EQ(tree->Root()->AncestorsAndSelf().size(), 1);
            EXPECT_EQ(tree->Root()->Activity(), Node::Activity_::Active);
            EXPECT_EQ(tree->Root()->Children().size(), 0);
            EXPECT_EQ(tree->Root()->Descendants().size(), 0);
            EXPECT_EQ(tree->Root()->DescendantsAndSelf().size(), 1);
        }
        {
            // tree.RemoveRoot
            tree->RemoveRoot(nullptr, [](auto *const root, [[maybe_unused]] const auto arg) { delete root; }); // NOLINT
            EXPECT_EQ(tree->Root(), nullptr);
        }

        delete tree;
    }

    TEST(Tests_00, Test_02) { // NOLINT
        auto *const root = new Root();
        auto *const a = new A();
        auto *const b = new B();

        for (const auto *const node : vector<Node *>{root, a, b}) {
            EXPECT_NE(node, nullptr);
            EXPECT_EQ(node->Tree(), nullptr);
            EXPECT_EQ(node->IsRoot(), true);
            EXPECT_EQ(node->Root(), node);
            EXPECT_EQ(node->Parent(), nullptr);
            EXPECT_EQ(node->Ancestors().size(), 0);
            EXPECT_EQ(node->AncestorsAndSelf().size(), 1);
            EXPECT_EQ(node->Activity(), Node::Activity_::Inactive);
            EXPECT_EQ(node->Children().size(), 0);
            EXPECT_EQ(node->Descendants().size(), 0);
            EXPECT_EQ(node->DescendantsAndSelf().size(), 1);
        }

        {
            // root.AddChildren
            root->AddChildren(vector<Node *>{a, b}, nullptr);
            EXPECT_NE(root, nullptr);
            EXPECT_EQ(root->Tree(), nullptr);
            EXPECT_EQ(root->IsRoot(), true);
            // EXPECT_EQ(root->Root(), root);
            EXPECT_EQ(root->Parent(), nullptr);
            EXPECT_EQ(root->Ancestors().size(), 0);
            EXPECT_EQ(root->AncestorsAndSelf().size(), 1);
            EXPECT_EQ(root->Activity(), Node::Activity_::Inactive);
            EXPECT_EQ(root->Children().size(), 2);
            EXPECT_EQ(root->Descendants().size(), 2);
            EXPECT_EQ(root->DescendantsAndSelf().size(), 3);
            for (const auto *const child : root->Children()) {
                EXPECT_EQ(child->Tree(), nullptr);
                EXPECT_EQ(child->IsRoot(), false);
                EXPECT_EQ(child->Root(), root);
                EXPECT_EQ(child->Parent(), root);
                EXPECT_EQ(child->Ancestors().size(), 1);
                EXPECT_EQ(child->AncestorsAndSelf().size(), 2);
                EXPECT_EQ(child->Activity(), Node::Activity_::Inactive);
                EXPECT_EQ(child->Children().size(), 0);
                EXPECT_EQ(child->Descendants().size(), 0);
                EXPECT_EQ(child->DescendantsAndSelf().size(), 1);
            }
        }
        {
            // root.RemoveChildren
            root->RemoveChildren(nullptr, [](auto *const child, [[maybe_unused]] const auto arg) { delete child; });
            EXPECT_NE(root, nullptr);
            EXPECT_EQ(root->Tree(), nullptr);
            EXPECT_EQ(root->IsRoot(), true);
            // EXPECT_EQ(root->Root(), root);
            EXPECT_EQ(root->Parent(), nullptr);
            EXPECT_EQ(root->Ancestors().size(), 0);
            EXPECT_EQ(root->AncestorsAndSelf().size(), 1);
            EXPECT_EQ(root->Activity(), Node::Activity_::Inactive);
            EXPECT_EQ(root->Children().size(), 0);
            EXPECT_EQ(root->Descendants().size(), 0);
            EXPECT_EQ(root->DescendantsAndSelf().size(), 1);
        }
    }

    TEST(Tests_00, Test_03) { // NOLINT
        auto *const tree = new Tree();
        auto *const root = new Root();
        auto *const a = new A();
        auto *const b = new B();

        for (const auto *const node : vector<Node *>{root, a, b}) {
            EXPECT_NE(node, nullptr);
            EXPECT_EQ(node->Tree(), nullptr);
            EXPECT_EQ(node->IsRoot(), true);
            EXPECT_EQ(node->Root(), node);
            EXPECT_EQ(node->Parent(), nullptr);
            EXPECT_EQ(node->Ancestors().size(), 0);
            EXPECT_EQ(node->AncestorsAndSelf().size(), 1);
            EXPECT_EQ(node->Activity(), Node::Activity_::Inactive);
            EXPECT_EQ(node->Children().size(), 0);
            EXPECT_EQ(node->Descendants().size(), 0);
            EXPECT_EQ(node->DescendantsAndSelf().size(), 1);
        }

        {
            // root.AddChildren a, b
            root->AddChildren(vector<Node *>{a, b}, nullptr);
        }
        {
            // tree.AddRoot root
            tree->AddRoot(root, nullptr);
            EXPECT_NE(tree->Root(), nullptr);
            EXPECT_EQ(tree->Root()->Tree(), tree);
            EXPECT_EQ(tree->Root()->IsRoot(), true);
            EXPECT_EQ(tree->Root()->Root(), root);
            EXPECT_EQ(tree->Root()->Parent(), nullptr);
            EXPECT_EQ(tree->Root()->Ancestors().size(), 0);
            EXPECT_EQ(tree->Root()->AncestorsAndSelf().size(), 1);
            EXPECT_EQ(tree->Root()->Activity(), Node::Activity_::Active);
            EXPECT_EQ(tree->Root()->Children().size(), 2);
            EXPECT_EQ(tree->Root()->Descendants().size(), 2);
            EXPECT_EQ(tree->Root()->DescendantsAndSelf().size(), 3);
        }
        {
            // tree.RemoveRoot
            tree->RemoveRoot(nullptr, [](auto *const root, [[maybe_unused]] const auto arg) { delete root; }); // NOLINT
            EXPECT_EQ(tree->Root(), nullptr);
        }

        delete tree;
    }

}

File: TreeMachine.Pro_CSharp/TreeMachine.Pro.Tests/System.TreeMachine/Tests_00.cs
namespace System.TreeMachine {
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;
    using NUnit.Framework;
    using Assert = NUnit.Framework.Assert;

    public class Tests_00 {

        [Test]
        public void Test_00() {
            var tree = new Tree();
            var root = new Root();
            var a = new A();
            var b = new B();
            {
                // tree.AddRoot root
                tree.AddRoot( root, null );
                Assert.That( tree.Root, Is.EqualTo( root ) );
                Assert.That( tree.Root.Tree, Is.EqualTo( tree ) );
                Assert.That( tree.Root.Tree_NoRecursive, Is.EqualTo( tree ) );
                Assert.That( tree.Root.IsRoot, Is.True );
                Assert.That( tree.Root.Root, Is.EqualTo( root ) );
                Assert.That( tree.Root.Parent, Is.Null );
                Assert.That( tree.Root.Ancestors.Count(), Is.EqualTo( 0 ) );
                Assert.That( tree.Root.AncestorsAndSelf.Count(), Is.EqualTo( 1 ) );
                Assert.That( tree.Root.Activity, Is.EqualTo( Node.Activity_.Active ) );
                Assert.That( tree.Root.Children.Count, Is.EqualTo( 0 ) );
                Assert.That( tree.Root.Descendants.Count, Is.EqualTo( 0 ) );
                Assert.That( tree.Root.DescendantsAndSelf.Count, Is.EqualTo( 1 ) );
            }
            {
                // tree.Root.AddChildren a, b
                tree.Root.AddChildren( [ a, b ], null );
                Assert.That( tree.Root, Is.EqualTo( root ) );
                Assert.That( tree.Root.Tree, Is.EqualTo( tree ) );
                Assert.That( tree.Root.Tree_NoRecursive, Is.EqualTo( tree ) );
                Assert.That( tree.Root.IsRoot, Is.True );
                Assert.That( tree.Root.Root, Is.EqualTo( root ) );
                Assert.That( tree.Root.Parent, Is.Null );
                Assert.That( tree.Root.Ancestors.Count(), Is.EqualTo( 0 ) );
                Assert.That( tree.Root.AncestorsAndSelf.Count(), Is.EqualTo( 1 ) );
                Assert.That( tree.Root.Activity, Is.EqualTo( Node.Activity_.Active ) );
                Assert.That( tree.Root.Children.Count, Is.EqualTo( 2 ) );
                Assert.That( tree.Root.Descendants.Count, Is.EqualTo( 2 ) );
                Assert.That( tree.Root.DescendantsAndSelf.Count, Is.EqualTo( 3 ) );
                foreach (var child in tree.Root.Children) {
                    Assert.That( child, Is.EqualTo( a ).Or.EqualTo( b ) );
                    Assert.That( child.Tree, Is.EqualTo( tree ) );
                    Assert.That( child.Tree_NoRecursive, Is.Null );
                    Assert.That( child.IsRoot, Is.False );
                    Assert.That( child.Root, Is.EqualTo( root ) );
                    Assert.That( child.Parent, Is.EqualTo( root ) );
                    Assert.That( child.Ancestors.Count(), Is.EqualTo( 1 ) );
                    Assert.That( child.AncestorsAndSelf.Count(), Is.EqualTo( 2 ) );
                    Assert.That( child.Activity, Is.EqualTo( Node.Activity_.Active ) );
                    Assert.That( child.Children.Count, Is.EqualTo( 0 ) );
                    Assert.That( child.Descendants.Count, Is.EqualTo( 0 ) );
                    Assert.That( child.DescendantsAndSelf.Count, Is.EqualTo( 1 ) );
                }
            }
            {
                // tree.Root.RemoveChildren a, b
                tree.Root.RemoveChildren( null, null );
                Assert.That( tree.Root.Children.Count, Is.Zero );
            }
            {
                // tree.RemoveRoot
                tree.RemoveRoot( null, null );
                Assert.That( tree.Root, Is.Null );
            }
        }

        [Test]
        public void Test_01() {
            var tree = new Tree();
            var root = new Root();
            var a = new A();
            var b = new B();
            {
                // tree.AddRoot root
                tree.AddRoot( root, null );
                Assert.That( tree.Root, Is.EqualTo( root ) );
                Assert.That( tree.Root.Tree, Is.EqualTo( tree ) );
                Assert.That( tree.Root.Tree_NoRecursive, Is.EqualTo( tree ) );
                Assert.That( tree.Root.IsRoot, Is.True );
                Assert.That( tree.Root.Root, Is.EqualTo( root ) );
                Assert.That( tree.Root.Parent, Is.Null );
                Assert.That( tree.Root.Ancestors.Count(), Is.EqualTo( 0 ) );
                Assert.That( tree.Root.AncestorsAndSelf.Count(), Is.EqualTo( 1 ) );
                Assert.That( tree.Root.Activity, Is.EqualTo( Node.Activity_.Active ) );
                Assert.That( tree.Root.Children.Count, Is.EqualTo( 0 ) );
                Assert.That( tree.Root.Descendants.Count, Is.EqualTo( 0 ) );
                Assert.That( tree.Root.DescendantsAndSelf.Count, Is.EqualTo( 1 ) );
            }
            {
                // tree.Root.AddChildren a, b
                tree.Root.AddChildren( [ a, b ], null );
                Assert.That( tree.Root, Is.EqualTo( root ) );
                Assert.That( tree.Root.Tree, Is.EqualTo( tree ) );
                Assert.That( tree.Root.Tree_NoRecursive, Is.EqualTo( tree ) );
                Assert.That( tree.Root.IsRoot, Is.True );
                Assert.That( tree.Root.Root, Is.EqualTo( root ) );
                Assert.That( tree.Root.Parent, Is.Null );
                Assert.That( tree.Root.Ancestors.Count(), Is.EqualTo( 0 ) );
                Assert.That( tree.Root.AncestorsAndSelf.Count(), Is.EqualTo( 1 ) );
                Assert.That( tree.Root.Activity, Is.EqualTo( Node.Activity_.Active ) );
                Assert.That( tree.Root.Children.Count, Is.EqualTo( 2 ) );
                Assert.That( tree.Root.Descendants.Count, Is.EqualTo( 2 ) );
                Assert.That( tree.Root.DescendantsAndSelf.Count, Is.EqualTo( 3 ) );
                foreach (var child in tree.Root.Children) {
                    Assert.That( child, Is.EqualTo( a ).Or.EqualTo( b ) );
                    Assert.That( child.Tree, Is.EqualTo( tree ) );
                    Assert.That( child.Tree_NoRecursive, Is.Null );
                    Assert.That( child.IsRoot, Is.False );
                    Assert.That( child.Root, Is.EqualTo( root ) );
                    Assert.That( child.Parent, Is.EqualTo( root ) );
                    Assert.That( child.Ancestors.Count(), Is.EqualTo( 1 ) );
                    Assert.That( child.AncestorsAndSelf.Count(), Is.EqualTo( 2 ) );
                    Assert.That( child.Activity, Is.EqualTo( Node.Activity_.Active ) );
                    Assert.That( child.Children.Count, Is.EqualTo( 0 ) );
                    Assert.That( child.Descendants.Count, Is.EqualTo( 0 ) );
                    Assert.That( child.DescendantsAndSelf.Count, Is.EqualTo( 1 ) );
                }
            }
            {
                // tree.RemoveRoot
                tree.RemoveRoot( null, null );
                Assert.That( tree.Root, Is.Null );
            }
        }

        [Test]
        public void Test_02() {
            var tree = new Tree();
            var root = new Root();
            var a = new A();
            var b = new B();
            {
                // root.AddChildren a, b
                root.AddChildren( [ a, b ], null );
                Assert.That( root, Is.EqualTo( root ) );
                Assert.That( root.Tree, Is.Null );
                Assert.That( root.Tree_NoRecursive, Is.Null );
                Assert.That( root.IsRoot, Is.True );
                Assert.That( root.Root, Is.EqualTo( root ) );
                Assert.That( root.Parent, Is.Null );
                Assert.That( root.Ancestors.Count(), Is.EqualTo( 0 ) );
                Assert.That( root.AncestorsAndSelf.Count(), Is.EqualTo( 1 ) );
                Assert.That( root.Activity, Is.EqualTo( Node.Activity_.Inactive ) );
                Assert.That( root.Children.Count, Is.EqualTo( 2 ) );
                Assert.That( root.Descendants.Count, Is.EqualTo( 2 ) );
                Assert.That( root.DescendantsAndSelf.Count, Is.EqualTo( 3 ) );
                foreach (var child in root.Children) {
                    Assert.That( child, Is.EqualTo( a ).Or.EqualTo( b ) );
                    Assert.That( child.Tree, Is.Null );
                    Assert.That( child.Tree_NoRecursive, Is.Null );
                    Assert.That( child.IsRoot, Is.False );
                    Assert.That( child.Root, Is.EqualTo( root ) );
                    Assert.That( child.Parent, Is.EqualTo( root ) );
                    Assert.That( child.Ancestors.Count(), Is.EqualTo( 1 ) );
                    Assert.That( child.AncestorsAndSelf.Count(), Is.EqualTo( 2 ) );
                    Assert.That( child.Activity, Is.EqualTo( Node.Activity_.Inactive ) );
                    Assert.That( child.Children.Count, Is.EqualTo( 0 ) );
                    Assert.That( child.Descendants.Count, Is.EqualTo( 0 ) );
                    Assert.That( child.DescendantsAndSelf.Count, Is.EqualTo( 1 ) );
                }
            }
            {
                // tree.AddRoot root
                tree.AddRoot( root, null );
                Assert.That( tree.Root, Is.EqualTo( root ) );
                Assert.That( tree.Root.Tree, Is.EqualTo( tree ) );
                Assert.That( tree.Root.Tree_NoRecursive, Is.EqualTo( tree ) );
                Assert.That( tree.Root.IsRoot, Is.True );
                Assert.That( tree.Root.Root, Is.EqualTo( root ) );
                Assert.That( tree.Root.Parent, Is.Null );
                Assert.That( tree.Root.Ancestors.Count(), Is.EqualTo( 0 ) );
                Assert.That( tree.Root.AncestorsAndSelf.Count(), Is.EqualTo( 1 ) );
                Assert.That( tree.Root.Activity, Is.EqualTo( Node.Activity_.Active ) );
                Assert.That( tree.Root.Children.Count, Is.EqualTo( 2 ) );
                Assert.That( tree.Root.Descendants.Count, Is.EqualTo( 2 ) );
                Assert.That( tree.Root.DescendantsAndSelf.Count, Is.EqualTo( 3 ) );
                foreach (var child in tree.Root.Children) {
                    Assert.That( child, Is.EqualTo( a ).Or.EqualTo( b ) );
                    Assert.That( child.Tree, Is.EqualTo( tree ) );
                    Assert.That( child.Tree_NoRecursive, Is.Null );
                    Assert.That( child.IsRoot, Is.False );
                    Assert.That( child.Root, Is.EqualTo( root ) );
                    Assert.That( child.Parent, Is.EqualTo( root ) );
                    Assert.That( child.Ancestors.Count(), Is.EqualTo( 1 ) );
                    Assert.That( child.AncestorsAndSelf.Count(), Is.EqualTo( 2 ) );
                    Assert.That( child.Activity, Is.EqualTo( Node.Activity_.Active ) );
                    Assert.That( child.Children.Count, Is.EqualTo( 0 ) );
                    Assert.That( child.Descendants.Count, Is.EqualTo( 0 ) );
                    Assert.That( child.DescendantsAndSelf.Count, Is.EqualTo( 1 ) );
                }
            }
            {
                // tree.RemoveRoot
                tree.RemoveRoot( null, null );
                Assert.That( tree.Root, Is.Null );
            }
        }

    }
}
